linfa-clustering/src/optics/algorithm.rs

```diff
@@ -0,0 +1,247 @@
+    fn set_core_distance(&mut self, min_pts: usize, neighbors: &[Neighbor<'a, F>]) {
+        self.c_distance = neighbors
+            .iter()
+            .nth(min_pts - 1)
+            .map(|x| FloatOrd(self.observation.l2_dist(&x.observation).unwrap()))
+            .clone();
+    }
+
+    /// Convert the neighbor to a sample for the user
+    fn sample(&self) -> Sample<'a, F> {
+        Sample {
+            index: self.index,
+            observation: self.observation.clone(),
+            reachability_distance: self.r_distance.map(|x| x.0),
+            core_distance: self.c_distance.map(|x| x.0),
+        }
+    }
+}
+
+impl<'a, F: Float> Eq for Neighbor<'a, F> {}
+
+impl<'a, F: Float> PartialEq for Neighbor<'a, F> {
+    fn eq(&self, other: &Self) -> bool {
+        self.r_distance == other.r_distance
+    }
+
+    fn ne(&self, other: &Self) -> bool {
+        !self.eq(other)
+    }
+}
+
+impl<'a, F: Float> PartialOrd for Neighbor<'a, F> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        self.r_distance.partial_cmp(&other.r_distance)
+    }
+}
+
+impl<'a, F: Float> Ord for Neighbor<'a, F> {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.r_distance.cmp(&other.r_distance)
+    }
+}
+
+impl Optics {
+    pub fn params(min_points: usize) -> OpticsHyperParamsBuilder {
+        OpticsHyperParams::new(min_points)
+    }
+}
+
+impl<'a, F: Float, D: Data<Elem = F>> Transformer<&'a ArrayBase<D, Ix2>, OpticsResult<'a, F>>
+    for OpticsHyperParams
+{
+    fn transform(&self, observations: &'a ArrayBase<D, Ix2>) -> OpticsResult<'a, F> {
+        let mut result = OpticsResult { orderings: vec![] };
+
+        let mut points = observations
+            .axis_iter(Axis(0))
+            .enumerate()
+            .map(|(i, x)| Neighbor::new(i, x))
+            .collect::<Vec<_>>();
+
+        let mut processed = HashSet::new();
+        while processed.len() != points.len() {
+            let obs = points
+                .iter_mut()
+                .enumerate()
+                .find(|x| !x.1.processed)
+                .unwrap();
+            if processed.contains(&obs.0) {
+                // We've processed this point so can move on
+                continue;
+            }
+            processed.insert(obs.0);
+            let neighbors = find_neighbors(&obs.1.observation, observations, self.tolerance());
+            let n = obs.1;
+            n.set_core_distance(self.minimum_points(), &neighbors);
+            result.orderings.push(n.sample());
+            if n.c_distance.is_some() {
+                // Here we get a list of "density reachable" samples that haven't been processed
+                // and sort them by reachability so we can process the closest ones first.
+                let mut seeds = Vec::new();
+                get_seeds(n.clone(), &neighbors, &mut points, &processed, &mut seeds);
+                while !seeds.is_empty() {
+                    seeds.sort_unstable_by(|a, b| b.cmp(a));
```

Comment:
`seeds` are the indices of points which have been added to a core the first time. It was unclear to me why ordering the indices also orders the distance in the core. Could you just add a sentence, which says that `get_neighbors` returns ordered element